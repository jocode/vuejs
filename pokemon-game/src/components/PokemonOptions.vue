<template>
  <div class="options-container">
    <ul>
      <li v-for="pokemon in pokemons" :key="pokemon.id"
      @click="selectPokemon(pokemon.id)" :class="{selected: (pokemon.id === selectedPokemon)}">
        {{ pokemon.name }}
      </li>
    </ul>
  </div>
</template>

<script>

export default {

  props: {
    pokemons: {
      type: Array,
      required: true,
    }
  },

  data(){
    return {
      selectedPokemon: null,
    }
  },

  methods: {
    selectPokemon(pokemonId) {
      if (this.selectedPokemon) return
      this.selectedPokemon = pokemonId
      this.$emit('selection', pokemonId)
    }
  },

}
</script>

<style scoped>
/* Pokemon Options */
ul {
  list-style-type: none;
}

.selected {
  background-color: rgba(0, 0, 0, 0.05);
}

li {
  background-color: white;
  border-radius: 5px;
  border: 1px solid rgba(0, 0, 0, 0.2);
  cursor: pointer;
  margin-bottom: 10px;
  width: 60vw;
  padding: .3em 0;
}

li:hover {
  background-color: rgba(0, 0, 0, 0.05);
}

.options-container {
  display: flex;
  justify-content: center;
}
</style>