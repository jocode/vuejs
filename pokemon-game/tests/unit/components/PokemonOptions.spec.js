import { shallowMount } from "@vue/test-utils"
import PokemonOptions from "@/components/PokemonOptions.vue"

import {pokemons} from "../mocks/pokemon.mock"

describe('PokemonOptions Component', () => {

    let wrapper 

    beforeEach(() => {
        wrapper = shallowMount(PokemonOptions, {
            props: {
                pokemons: pokemons
            }
        })
    })

    test('debe de hacer match con el snapshot', () => {
        console.log(wrapper.html())
        expect(wrapper.html()).toMatchSnapshot()
        // expect(wrapper.html()).toMatchInlineSnapshot()
    })

    test('debe de mostrar las 4 opciones correctamente', () => {
        const liTags = wrapper.findAll('li')
        expect(liTags.length).toBe(4)
        expect(liTags[0].text()).toBe('pikachu')
        expect(liTags[1].text()).toBe('charmander')
        expect(liTags[2].text()).toBe('bulbasaur')
        expect(liTags[3].text()).toBe('squirtle')
    })

    test('debe de emitir "selection" con sus respectivos parámetros al hacer click', () => {
        const liTags = wrapper.findAll('li')
        liTags[0].trigger('click')
        // console.log(wrapper.emitted('selection'))

        /**
         * Es este caso no se hacen más emisiones, porque se verifica que sólo se haya hecho click una sola vez
         */

        expect(wrapper.emitted().selection).toBeTruthy()
        expect( wrapper.emitted('selection').length ).toBe(1)
        expect(wrapper.emitted().selection[0]).toEqual([5])
    })

})