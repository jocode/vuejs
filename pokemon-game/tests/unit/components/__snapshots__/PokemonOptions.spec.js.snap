// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonOptions Component debe de hacer match con el snapshot 1`] = `
<div class="options-container">
  <ul>
    <li class="">pikachu</li>
    <li class="">charmander</li>
    <li class="">bulbasaur</li>
    <li class="">squirtle</li>
  </ul>
</div>
`;
